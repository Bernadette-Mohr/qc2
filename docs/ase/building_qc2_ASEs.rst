.. _build_ASEs:

Building your own qc2-ASE calculator
====================================
