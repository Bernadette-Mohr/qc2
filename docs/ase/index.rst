ASE calculators
===============


.. toctree::
  :maxdepth: 1
