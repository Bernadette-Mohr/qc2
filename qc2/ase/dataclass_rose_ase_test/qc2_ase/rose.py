"""This module defines an ASE interface to Rose.

Original paper & official release:
https://pubs.acs.org/doi/10.1021/acs.jctc.0c00964

GitLab page:
https://gitlab.com/quantum_rose/rose

Note: see also https://pubs.acs.org/doi/10.1021/ct400687b.
"""
from dataclasses import dataclass, field
from enum import Enum
from ase import Atoms
from ase.calculators.calculator import FileIOCalculator
from typing import Optional, List, Union, Sequence


class RoseCalcType(Enum):
    """Enumerator class defining the type of calculation done by Rose."""
    ATOM_FRAG = 'atom_frag'  # calc IAOs from atom-free AOs.
    MOL_FRAG = 'mol_frag'    # calc IFOs from free molecular fragment MOs.


class RoseInterfaceMOCalculators(Enum):
    """Enumerator class defining the program to which Rose is interfaced."""
    PYSCF = 'pyscf'
    PSI4 = 'psi4'
    DIRAC = 'dirac'
    ADF = 'adf'
    GAUSSIAN = 'gaussian'


@dataclass
class RoseInputDataClass:
    """A dataclass representing input options for Rose."""
    rose_target: Atoms
    rose_frags: Union[Atoms, Sequence[Atoms]]

    calculate_mo: bool = True
    rose_calc_type: RoseCalcType = RoseCalcType.ATOM_FRAG.value

    run_postscf: bool = False
    restricted: bool = True
    openshell: bool = False
    relativistic: bool = False

    # Rose intrinsic options => expected to be fixed at the default values?
    version: str = ['Stndrd_2013', 'Simple_2013', 'Simple_2014'][0]
    exponent: int = [2, 3, 4][2]
    spherical: bool = False
    uncontract: bool = True
    test: bool = True
    wf_restart: bool = True
    get_oeint: bool = True
    save: bool = True
    avas_frag: Optional[List[int]] = field(default_factory=list)
    nmo_avas: Optional[List[int]] = field(default_factory=list)

    def __post_init__(self):
        """This method is called after the instance is initialized."""
        if self.test:
            self.get_oeint = True

    """Description of the attributes to be added below."""


class Rose(RoseInputDataClass, FileIOCalculator):
    """A general ASE calculator for ROSE (Reduction of Orbital Space Extent).

    Args:
        RoseInputDataClass (RoseInputDataClass): Tailor-made base dataclass
            for Rose; see above.
        FileIOCalculator (FileIOCalculator): Base class for ASE calculators
            that write/read input/output files.

    """
    implemented_properties: List[str] = []
    command: str = 'echo "Executing Rose...done"'
    restart: bool = None
    ignore_bad_restart_file: bool = FileIOCalculator._deprecated
    label: str = 'rose'
    atoms: Atoms = None

    def __init__(self, *args, **kwargs) -> None:
        """ASE-Rose Class Constructor to initialize the object.

        Give an example here on how to use....
        """
        # initializing base classes
        RoseInputDataClass.__init__(self, *args, **kwargs)
        FileIOCalculator.__init__(self, *args, **kwargs)

    def calculate(self) -> None:
        """Executes Rose workflow."""
        self.get_input_genibo_avas()
        self.get_input_xyz()

        if self.calculate_mo:
            self.get_mo_files()

        self.run_rose()

        if self.save:
            self.save_ibos()

        if len(self.avas_frag) != 0:
            self.run_avas()

        if self.run_postscf:
            self.run_post_hf()

    def get_input_genibo_avas(self) -> None:
        """Generates INPUT_GENIBO & INPUT_AVAS fortran files for Rose."""
        write_input_genibo_avas(self)

    def get_input_xyz(self) -> None:
        """Generates Molecule and Fragment xyz files for Rose."""
        print("Creating Molecule and Frags inputs....done")

    def get_mo_files(self) -> None:
        """Generates atomic and molecular orbitals files for Rose."""
        print('Calculating MO files....done')

    def run_rose(self) -> None:
        """Runs Rose executable 'genibo.x'."""
        print('Executing Rose....done')

    def save_ibos(self) -> None:
        """Generates a checkpoint file ('ibo.chk') with the final IBOs."""
        print('Saving ibos....done')

    def run_avas(self) -> None:
        """Runs 'avas.x' executable."""
        print('Executing avas....done')

    def run_post_hf(self) -> None:
        """Performs CASCI and/or CASSCF calculations."""
        print('CASSCF?....done')


def write_input_genibo_avas(input_data: RoseInputDataClass) -> None:
    """Summary.

    Args:
        input_data (RoseInputDataClass): description
    """
    print("Creating genibo and avas input....done")

    molecule_charge = input_data.rose_target.calc.parameters.charge
    molecule_mo_calculator = input_data.rose_target.calc.name.lower()
    molecule_natom = len(input_data.rose_target.symbols)

    fragments_mo_calculator = []
    for n in range(len(input_data.rose_frags)):
        frag_mo_program = input_data.rose_frags[n].calc.name.lower()
        fragments_mo_calculator.append(frag_mo_program)

    version = input_data.version
    exponent = input_data.exponent
    restricted = input_data.restricted
    test = input_data.test

    avas_frag = input_data.avas_frag
    nmo_avas = input_data.nmo_avas

    # creating INPUT_GENIBO file
    with open("INPUT_GENIBO", "w") as f:
        f.write("**ROSE\n")
        f.write(".VERSION\n")
        f.write(version+"\n")
        f.write(".CHARGE\n")
        f.write(str(molecule_charge)+"\n")
        f.write(".EXPONENT\n")
        f.write(str(exponent)+"\n")
        if not restricted:
            f.write(".UNRESTRICTED\n")
        f.write(".FILE_FORMAT\n")
        f.write(molecule_mo_calculator+"\n")
        if test == 1:
            f.write(".TEST  \n")
        f.write(".AVAS  \n")
        f.write(str(len(avas_frag))+"\n")
        f.writelines("{:3d}".format(item) for item in avas_frag)
        f.write("\n*END OF INPUT\n")

    # creating INPUT_AVAS file
    with open("INPUT_AVAS", "w") as f:
        f.write(str(molecule_natom) + " # natoms\n")
        f.write(str(molecule_charge) + " # charge\n")
        if restricted:
            f.write("1 # restricted\n")
        else:
            f.write("0 # restricted\n")
        f.write("1 # spatial orbs\n")
        f.write(molecule_mo_calculator + " # MO file for the full molecule\n")
        f.write(fragments_mo_calculator + " # MO file for the fragments\n")
        f.write(str(len(nmo_avas)) + " # number of valence MOs in B2\n")
        f.writelines("{:3d}".format(item) for item in nmo_avas)

    # print(molecule_natom)
