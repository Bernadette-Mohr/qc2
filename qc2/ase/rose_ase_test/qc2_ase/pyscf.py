"""This module defines an ASE interface to PySCF.

https://pyscf.org/ => Official documentation
https://github.com/pyscf/pyscf => GitHub page

Note: Adapted from
https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/tools/pyscf_ase.py
&
https://github.com/pyscf/pyscf/issues/624
"""
from typing import Union, Optional, List, Dict, Any
import warnings
import numpy as np
from ase import Atoms
from ase.calculators.calculator import Calculator, all_changes
from ase.units import Ha, Bohr
from ase.calculators.calculator import InputError
from ase.calculators.calculator import CalculatorSetupError


def ase_atoms_to_pyscf(ase_atoms: Atoms) -> List[List[Union[str, np.ndarray]]]:
    """Converts ASE atoms to PySCF atom.

    Args:
        ase_atoms (Atoms): ASE Atoms object.

    Returns:
        List[List[Union[str, np.ndarray]]]: PySCF atom.
    """
    return [[ase_atoms.get_chemical_symbols()[i], ase_atoms.get_positions()[i]]
            for i in range(len(ase_atoms.get_positions()))]


class PySCF(Calculator):
    """An ASE calculator for the PySCF quantum chemistry package.

    Args:
        Calculator (Calculator): Base-class for all ASE calculators.

    Raises:
        InputError: If attributes other than
            'method', 'xc', 'basis', 'multiplicity',
            'charge', and 'verbose' are input as Calculator.
        CalculatorSetupError: If abinitio methods other than
            'scf.RHF', 'scf.UHF', 'scf.ROHF',
            'dft.RKS', 'dft.UKS', and 'dft.ROKS' are selected.

    Example of a typical ASE-PySCF input:

    >>> from ase import Atoms
    >>> from ase.build import molecule
    >>> from qc2.ase.pyscf import PySCF
    >>>
    >>> molecule = Atoms(...) or molecule = molecule('...')
    >>> molecule.calc = PySCF(method='dft.RKS',
                            xc='b3lyp',
                            basis='6-31g*',
                            charge=0,
                            multiplicity=1,
                            verbose=0)
    >>> energy = molecule.get_potential_energy()
    >>> gradient = molecule.get_forces()

    where (currently)

    method = 'scf.RHF'
           = 'scf.UHF'
           = 'scf.ROHF'
           = 'dft.RKS'
           = 'dft.UKS'
           = 'dft.ROKS'
    """
    implemented_properties: List[str] = ['energy', 'forces']

    default_parameters: Dict[str, Any] = {'method': 'dft.RKS',
                                          'basis': '6-31g*',
                                          'xc': 'b3lyp',
                                          'multiplicity': 1,
                                          'charge': 0,
                                          'verbose': 0}

    def __init__(self,
                 restart: Optional[bool] = None,
                 ignore_bad_restart: Optional[bool] = False,
                 label: Optional[str] = 'PySCF',
                 atoms: Optional[Atoms] = None,
                 command: Optional[str] = None,
                 directory: str = '.',
                 **kwargs) -> None:
        """ASE-PySCF Class Constructor to initialize the object.

        Note: Basic implementation based on the
            class Psi4(Calculator); see, e.g., ase/ase/calculators/psi4.py.

        Args:
            restart (bool, optional): Prefix for restart file.
                May contain a directory. Defaults to None: don't restart.
            ignore_bad_restart (bool, optional): Deprecated and will
                stop working in the future. Defaults to False.
            label (str, optional): Calculator name. Defaults to 'PySCF'.
            atoms (Atoms, optional): Atoms object to which the calculator
                will be attached. When restarting, atoms will get its
                positions and unit-cell updated from file. Defaults to None.
            command (str, optional): Command used to start calculation.
                Defaults to None.
            directory (str, optional): Working directory in which
                to perform calculations. Defaults to '.'.
        """
        # initializing base class Calculator.
        # see ase/ase/calculators/calculator.py.
        super().__init__(restart=restart,
                         ignore_bad_restart=ignore_bad_restart, label=label,
                         atoms=atoms, command=command, directory=directory,
                         **kwargs)
        """Transforms **kwargs into a dictionary with calculation parameters.

        Starting with (attr1=value1, attr2=value2, ...)
            it creates self.parameters['attr1']=value1, and so on.
        """

        # Check self.parameters input keys and values
        self.check_pyscf_attributes()

    def check_pyscf_attributes(self) -> None:
        """Checks for any missing and/or mispelling PySCF input attribute.

        Note: it can also be used to eventually set specific
        environment variables, ios, etc.
        """
        recognized_attributes: List[str] = [
            'ignore_bad_restart', 'command', 'method',
            'xc', 'basis', 'multiplicity', 'charge',
            'verbose', 'kpts', 'nbands', 'smearing'
            ]

        # self.parameters gathers all PYSCF input options in a dictionary.
        # it is defined by class Calculator(BaseCalculator)
        # see ase/ase/calculators/calculator.py
        for key, value in self.parameters.items():
            # check attribute name
            if key not in recognized_attributes:
                raise InputError('Attribute', key,
                                 ' not recognized. Please check input.')

        # dealing with lack of multiplicity and charge info
        if 'multiplicity' not in self.parameters.keys():
            self.parameters['multiplicity'] = 1
            warnings.warn('Multiplicity not provided.'
                          'Assuming default singlet.')

        if 'charge' not in self.parameters.keys():
            self.parameters['charge'] = 0
            warnings.warn('Charge not provided. Assuming default zero.')

        # verbose sets the amount of pyscf printing
        # verbose = 0 prints no info
        if 'verbose' not in self.parameters.keys():
            self.parameters['verbose'] = 0

        # dealing with some ASE specific inputs
        if 'kpts' in self.parameters.keys():
            raise InputError('This ASE-PySCF interface does not yet implement'
                             ' periodic calculations, and thus does not'
                             ' accept k-points as parameters. '
                             'Please remove this argument.')

        if 'nbands' in self.parameters.keys():
            raise InputError('This ASE-PySCF interface '
                             'does not support the keyword "nbands".')

        if 'smearing' in self.parameters.keys():
            raise InputError('Finite temperature DFT is not currently'
                             ' implemented in this PySCF-ASE interface,'
                             ' thus a smearing argument cannot be utilized.'
                             ' Please remove this argument.')

    def calculate(self, atoms: Optional[Atoms] = None,
                  properties: List[str] = ['energy'],
                  system_changes: List[str] = all_changes) -> None:
        """This method is the core responsible for the actual calculation.

        Note: Implementation based on Calculator.calculate() method.
            see also ase/ase/calculators/calculator.py.

        Args:
            atoms (Atoms, optional): Atoms object to which
                the calculator is attached. Defaults to None.
            properties (list[str], optional): List of what properties
                need to be calculated. Defaults to ['energy'].
            system_changes (list[str], optional): List of what has changed
                since last calculation. Can be any combination
                of these six: 'positions', 'numbers', 'cell', 'pbc',
                'initial_charges' and 'initial_magmoms'.
                Defaults to all_changes.

        Raises:
            CalculatorSetupError: If a proper geometry is not provided.
            CalculatorSetupError: If abinitio methods other than
            'scf.RHF', 'scf.UHF', 'scf.ROHF', 'dft.RKS', 'dft.UKS',
                and 'dft.ROKS' are selected.
        """
        implemented_methods: List[str] = ['scf.RHF', 'scf.UHF', 'scf.ROHF',
                                          'dft.RKS', 'dft.UKS', 'dft.ROKS']

        from pyscf import gto  # , scf, dft

        # setting up self.atoms attribute from base class Calculator.calculate.
        # this is extracted from the atoms Atoms object.
        super().calculate(atoms=atoms)

        # checking that self.atoms has been properly initiated/updated
        if self.atoms is None:
            raise CalculatorSetupError('An Atoms object must be provided to '
                                       'perform a calculation.')

        # the spin variable corresponds to 2S instead of 2S+1
        spin_2s = self.parameters['multiplicity'] - 1

        # passing geometry and other definitions
        molecule = gto.M(atom=ase_atoms_to_pyscf(self.atoms),
                         basis=self.parameters['basis'],
                         charge=self.parameters['charge'],
                         spin=spin_2s)

        # checking wf input name => this is case sensitive
        if self.parameters['method'] in implemented_methods:
            self.wf = eval(self.parameters['method'])(molecule)
        else:
            raise CalculatorSetupError('Method not yet implemented. '
                                       'Current PySCF-ASE calculator '
                                       'only allows for',
                                       implemented_methods,
                                       'wave functions.'
                                       ' Please check input method.')

        self.wf.verbose = self.parameters['verbose']

        if 'dft' in self.parameters['method']:
            self.wf.xc = self.parameters['xc']

        # calculating energy in eV
        energy = self.wf.kernel() * Ha
        self.results['energy'] = energy

        # calculating forces
        if 'forces' in properties:
            gf = self.wf.nuc_grad_method()
            gf.verbose = self.parameters['verbose']
            if 'dft' in self.parameters['method']:
                gf.grid_response = True
            # analytic gradienta in eV/Angstrom
            forces = -1.0 * gf.kernel() * (Ha / Bohr)
            totalforces = []
            totalforces.extend(forces)
            totalforces = np.array(totalforces)
            self.results['forces'] = totalforces
