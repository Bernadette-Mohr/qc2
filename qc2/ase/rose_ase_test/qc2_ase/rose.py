"""This module defines an ASE interface to ROSE.

Original paper & official release:
https://pubs.acs.org/doi/10.1021/acs.jctc.0c00964

GitLab page:
https://gitlab.com/quantum_rose/rose

Note: see also https://pubs.acs.org/doi/10.1021/ct400687b.
"""
from ase import Atoms
from ase.calculators.calculator import FileIOCalculator
from ase.units import Ha  # => needed only for testing; remove latter.
from typing import Optional  # List, Tuple, Union
# from qc2_ase.rose_io import *
# from qc2_ase.pyscf import PySCF


class ROSE(FileIOCalculator):
    """A general ASE calculator for ROSE (Reduction of Orbital Space Extent).

    Args:
        FileIOCalculator (FileIOCalculator): Base class for calculators
            that write/read input/output files.
    """
    #implemented_properties = ['orbitals']
    command = 'echo "Executing Rose...done"'  # => test

    default_parameters = {
        'rose_calc_type': ['atom_frag', 'mol_frag'][1],
        'calculate_mo': True,
        'uncontract': True,
        'version': ['Stndrd_2013', 'Simple_2013', 'Simple_2014'][0],
        'exponent': [2, 3, 4][2],
        'relativistic': False,
        'spherical': False
        }

    def __init__(self,
                 restart: Optional[bool] = None,
                 ignore_bad_restart_file:
                 Optional[bool] = FileIOCalculator._deprecated,
                 label: Optional[str] = 'rose',
                 atoms: Optional[Atoms] = None,
                 command: Optional[str] = None,
                 **kwargs) -> None:
        """ASE-Rose Class Constructor to initialize the object."""
        super().__init__(restart, ignore_bad_restart_file,
                         label, atoms, command, **kwargs)
        """Transforms **kwargs into a dictionary with calculation parameters.

        Starting with (attr1=value1, attr2=value2, ...)
            it creates self.parameters['attr1']=value1, and so on.
        """

        print(self.parameters)
        # print(self.parameters.rose_target.calc)
        # print(self.parameters.rose_frags[0].calc)
        # print(self.parameters.rose_frags[1].calc)

    def calculate(self, *args, **kwargs):
        """Executes Rose workflow."""
        super().calculate(*args, **kwargs)
        # calls:
        # 1) write_input()
        # 2) execute()
        # 3) read_results()

    def write_input(self, atoms, properties=None, system_changes=None):
        """Generates all inputs necessary for Rose."""
        super().write_input(atoms, properties, system_changes)
        #
        # TODO
        #
        # 1) Write INPUT_GENIBO.
        # 2) Generate Molecule.xyz and Frags.xyz.
        # 3) If requested, calculate "on-the-fly" the orbitals file
        #       for the system and all its fragments

        print("Writing INPUT_GENIBO and INPUT_AVAS....done\n")

        print("Generating Molecule.xyz and Frags.xyz....done\n")

        print("Generating orbitals file....done")

        # test
        mol = self.parameters.rose_target
        print('H2O energy/Eh =', mol.get_potential_energy()/Ha)
        print('H2O orbitals =', mol.calc.wf.mo_coeff)

        for fragment in self.parameters.rose_frags:
            print(fragment.symbols, 'energy/Eh =',
                  fragment.get_potential_energy()/Ha)
            print(fragment.calc.wf.mo_coeff)
        print(" ")

    def execute(self):
        """_summary_."""
        super().execute()
        #
        # TODO
        #
        # 1) Execute genibo.x.
        # 2) Store IAOs & IBOs.
        # 3) Execute avas.x, if required.
        # 4) Post-HF calculations, if required.

    # def read_results(self):
    #    """_summary_"""
    #    #
    #    # TODO
    #    #
    #    # 1) Print SCF and/or post-HF energies
    #    self.results['energy'] = 1.0
    #    #print(self.results)
    #    print("Reading/Printing results...done")
