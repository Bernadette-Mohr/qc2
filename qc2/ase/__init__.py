"""qc2 ASE package."""
from .pyscf import PySCF
from .rose import ROSE, ROSETargetMolecule, ROSEFragment
from .dirac import DIRAC
