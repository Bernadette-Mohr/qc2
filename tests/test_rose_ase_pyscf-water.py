import os
import pytest

from .rose_test_functions import clean_up, extract_number, read_output

from ase import Atoms
from qc2.ase.rose import Rose
from qc2.ase.pyscf import PySCF


def run_water_rose_no_avas():
    """Water atom_frag example calculation."""
    # define target molecule
    mol = Atoms('OH2',
            positions=[[0.,  0.00000,  0.59372],
                       [0.,  0.76544, -0.00836],
                       [0., -0.76544, -0.00836]
                       ],
               calculator=PySCF(method='scf.RHF',
                                basis='unc-sto-3g',
                                multiplicity=1,
                                )
               )

    # define atomic fragments
    frag1 = Atoms('O',
              calculator=PySCF(method='scf.RHF',
                               basis='unc-sto-3g',
                               multiplicity=1,
                               )
              )

    frag2 = Atoms('H',
              calculator=PySCF(method='scf.ROHF',
                               basis='unc-sto-3g',
                               multiplicity=2
                               )
              )

    # perform scf calculations on all species
    mol.get_potential_energy()
    frag1.get_potential_energy()
    frag2.get_potential_energy()

    # prepare Rose-ASE calculator
    rose_calc = Rose(rose_calc_type='atom_frag',
                 rose_target=mol,
                 rose_frags=[frag1, frag2],
                 test = True
                 )

    # run the calculator
    rose_calc.calculate()


EXPECTED_OUTPUT_FILE = 'test_rose_ase_pyscf-water.stdout'
ACTUAL_OUTPUT_FILE = 'OUTPUT_ROSE'

CHARGE_REGEX = 'Partial charges of fragments.*\n{}'.format(('*\n.' * 6) + '*\n')
MO_ENERGIES_REGEX = 'Recanonicalized virtual energies of fragments.*\n{}'.format(('*\n.' * 9) + '*\n')
HF_ENERGIES_REGEX = 'HF energy .*\n{}'.format(('*\n.' * 2) + '*\n')


def run_calculation():
    """Calculation to run."""
    run_water_rose_no_avas()


@pytest.fixture(scope="session", autouse=True)
def clean_up_files():
    """Runs always at the end of each test."""
    yield
    clean_up()


def test_partial_charges():
    """Test case #1 - Partial charges of fragments."""
    # run the calculation
    run_calculation()

    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_charges = extract_number(CHARGE_REGEX, expected_output)
    actual_charges = extract_number(CHARGE_REGEX, actual_output)
    assert actual_charges == pytest.approx(expected_charges, rel=1e-3)


def test_virtual_energies():
    """Test case #2 - Recanonicalized virtual energies of fragments."""
    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_energies = extract_number(MO_ENERGIES_REGEX, expected_output)
    actual_energies = extract_number(MO_ENERGIES_REGEX, actual_output)
    assert actual_energies == pytest.approx(expected_energies, rel=1.0e-5)


def test_hf_energy():
    """Test case #3 - Check final HF energies, if test = True."""
    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_energy = extract_number(HF_ENERGIES_REGEX, expected_output)
    actual_energy = extract_number(HF_ENERGIES_REGEX, actual_output)
    assert actual_energy == pytest.approx(expected_energy, rel=1.0e-8)
