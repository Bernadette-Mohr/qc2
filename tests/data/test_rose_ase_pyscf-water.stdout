
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                     (ROSE)                     @
 @                                                @
 @ By Bruno Senjean, Souloke Sen, Michal Repisky, @
 @        Gerald Knizia and Lucas Visscher.       @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

 Select algebra (rcq) based on the use of restricted and spatial orbitals:
   rcq = 1, 2 or 4  : real, complex or quaternion algebra, respectively
   restricted       :  T
   spatial_orbitals :  T
   rcq              :  1

 ####################################
 #      Molecular  Coordinates      #
 ####################################

Atom    Z           X           Y           Z
O    8.00  0.00000000  0.00000000  1.12196819
H    1.00  0.00000000  1.44647196 -0.01579811
H    1.00  0.00000000 -1.44647196 -0.01579811

 ####################################
 #         Basis set size           #
 #               and                #
 #      Canonical MO Energies       #
 ####################################

 No. basis functions    B1:     21
 No. molecular orbitals B1:     20
 No. occupied  orbitals B1:      4
 No. molecular orbitals B2:      6


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.1151E-14


  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.3537E-14

 ####################################
 #  Partial charges of fragments    #
 ####################################

       Fragment      Charge
   O              -0.550108
   H               0.275054
   H               0.275054



 ####################################
 # Intrinsic Localized occupied MOs #
 ####################################


 ----- Percentage of localization before the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -1.3179 [    1    2    3]
                [ 73.4 13.3 13.3]
  2     -0.6598 [    1    2    3]
                [ 63.0 18.5 18.5]
  3     -0.5175 [    1    2    3]
                [ 91.1  4.4  4.4]
  4     -0.4560 [    1]
                [100.0]

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        2.140506    3.73E-02
    2        2.335013    2.64E-04
    3        2.364897    3.95E-06
    4        2.364897    1.74E-09
    5        2.364897    7.64E-13
    6        2.364897    3.41E-16

 ******************************************
 Final converged values (it, func, fGrad):
    6            2.3648967767875022  3.4120E-16
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1    3]
    [ 63.8 36.2]
  2 [    1    2]
    [ 63.8 36.2]
  3 [    1]
    [100.0]
  4 [    1]
    [100.0]

 ----- Percentage of localization after recanonization -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -1.3179 [    1    2    3]
                [ 73.4 13.3 13.3]
  2     -0.6598 [    1    2    3]
                [ 63.0 18.5 18.5]
  3     -0.5175 [    1    2    3]
                [ 91.1  4.4  4.4]
  4     -0.4560 [    1]
                [100.0]


 ############################################
 # Intrinsic Localized valence virtual MOs  #
 ############################################


 ----- Percentage of localization before the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1      0.2524 [    1    2    3]
                [ 35.5 32.3 32.3]
  2      0.3512 [    1    2    3]
                [ 37.0 31.5 31.5]

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        0.075958    4.78E-16

 ******************************************
 Final converged values (it, func, fGrad):
    1            0.0759576827009510  4.7786E-16
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1    2]
    [ 36.2 63.8]
  2 [    1    3]
    [ 36.2 63.8]

 ####################################
 #      Additional virtuals         #
 ####################################

 No. molecular virtuals B3:     14
 No. selected  virtuals B3:      5
 No. reference virtuals B4:     11


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.2849E-14


  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.6887E-14


 ##############################################
 # Intrinsic Localized additional virtual MOs #
 ##############################################


 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        3.172985    6.73E-02
    2        4.823797    1.60E-03
    3        4.823832    7.70E-07
    4        4.823832    3.11E-10
    5        4.823832    2.22E-15
    6        4.823832    8.84E-17

 ******************************************
 Final converged values (it, func, fGrad):
    6            4.8238320453795689  8.8375E-17
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1]
    [100.0]
  2 [    1]
    [ 99.5]
  3 [    1]
    [ 99.9]
  4 [    1    2]
    [  1.9 98.0]
  5 [    1    3]
    [  1.9 98.0]

 ----- Recanonicalized virtual energies of fragments -----

    Fragment          MO              Energy
           1           1         1.543099244
           1           2         1.597453330
           1           3         1.683789289
           2           1         0.343793188
           2           2         0.947674604
           3           1         0.343793188
           3           2         0.947674604

 ####################################
 #         Starting TESTS           #
 ####################################

 The matrix is normalized.
 The matrix is normalized.
 The matrix is normalized.
 Deviation of IBO_RDM from HF_RDM  :   1.0756043840582849E-014
 Deviation from orthogonality      :   3.6208046498591820E-015
 Deviation from orthogonality      :   5.1121204255647359E-015
 Deviation in number of electrons  :   8.8817841970012523E-015
 Deviation in number of electrons  :   8.8817841970012523E-015
 HF energy (moB1)   :   -75.162062930326002
 HF energy (IAO)    :   -75.162062930326073
 HF energy (IBO/MO) :   -75.162062930326044
 Deviation from the HF energy in the original basis and in the IAO basis:  -7.10542736E-14
 Deviation from the HF energy in the original basis and in the IBO basis:  -4.26325641E-14

 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                                                @
 @                      END                       @
 @                                                @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
